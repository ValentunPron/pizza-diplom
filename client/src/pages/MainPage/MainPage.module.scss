.main {
   max-width: 1340px;
   width: 100%;
   display: flex;
   flex-direction: column;
}

.products {
   width: 100%;
   display: flex;
   flex-direction: column;
   background-color: #fff;
   padding: 40px 15px;
   border-radius: 13px;
   .productsActions {
      width: 100%;
      display: flex;
      justify-content: space-between;
      align-items: center;
      margin-bottom: 30px;
   }
   .productsTitle {
      font-size: 28px;
      background-color: transparent;
      font-family: "Proxima-Nova";
      font-weight: 400;
      margin-bottom: 35px;
   }
   .productsContent {
      width: 100%;
      background-color: #fff;
      display: grid;
      grid-template-columns: repeat(auto-fill, 290px);
      grid-auto-flow: row;
      justify-items: center;
      justify-content: center;
      gap: 50px 39px;
   }
}

@media (max-width: 1080px) {
   .products {
      .productsActions {
         flex-direction: column;
         margin-bottom: 50px;
      }
   }
}
