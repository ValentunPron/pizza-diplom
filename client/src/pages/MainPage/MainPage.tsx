import { FC, useState } from 'react';

import s from './MainPage.module.scss';
import { useAppSelector } from 'hooks/redux';
import { IPizza } from 'types/pizza';

import PizzaItem from 'components/MainPage/PizzaItem/PizzaItem';
import Filter from 'components/MainPage/Filter/Filter';
import Loader from 'components/Common/Loader/Loader';
import Sorter from 'components/MainPage/Sorter/Sorter';

const MainPage: FC = () => {

   const { status } = useAppSelector(state => state.productsSlice);
   // капия исходного списка пицц, чтобы делать различные манипуляции с сортировками
   const [sortedPizza, setSortedPizza] = useState<IPizza[]>([]);

   return (
      <div className={s.main} data-testid="main-page">
         <div className={s.products}>

            {status === 'pending' && <Loader />}

            {status === 'ready' &&
               <>
                  <div className={s.productsActions}>
                     <Filter setSortedPizza={setSortedPizza} />

                     <Sorter setSortedPizza={setSortedPizza} />
                  </div>

                  <p className={s.productsTitle}>Пицца</p>

                  <div className={s.productsContent}>
                     {sortedPizza.map(item => {
                        return <PizzaItem pizza={item} key={item.name} />;
                     })}
                  </div>
               </>
            }
         </div>
      </div>
   );
};

export default MainPage;