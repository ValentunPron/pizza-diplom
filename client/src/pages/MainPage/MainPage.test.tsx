import { fireEvent, render, screen } from '@testing-library/react';
import selectEvent from 'react-select-event';

import { renderWithReduxAndRouter } from 'tests/helpers/renderWithReduxAndRouter';
import { productsSlice } from "tests/mocks/productsSlice";

import MainPage from "./MainPage";

describe('MainPage products functional', () => {
   it('must be 10 pizzas in page', () => {
      render(renderWithReduxAndRouter(<MainPage />, ['/'], { productsSlice }));

      expect(screen.getAllByTestId(/pizza-item/i)).toHaveLength(10);
   });

   it('checking filter operability', () => {
      render(renderWithReduxAndRouter(<MainPage />, ['/'], { productsSlice }));

      const meatFilterBtn = screen.getByRole('button', { name: /мясные/i });
      fireEvent.click(meatFilterBtn);
      expect(screen.getAllByTestId(/pizza-item/i)).toHaveLength(6);

      const veganFilterBtn = screen.getByRole('button', { name: /вегетарианские/i });
      fireEvent.click(veganFilterBtn);
      expect(screen.getAllByTestId(/pizza-item/i)).toHaveLength(4);
   });

   it('checking sorter operability', async () => {
      const { container } = render(renderWithReduxAndRouter(<MainPage />, ['/'], { productsSlice }));

      await selectEvent.select(screen.getByText('популярности'), ['цене (макс)']);
      expect(container).toMatchSnapshot();
   });
});