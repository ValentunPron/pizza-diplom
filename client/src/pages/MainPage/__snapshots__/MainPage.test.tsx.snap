// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage products functional checking sorter operability 1`] = `
<div>
  <div
    class="main"
    data-testid="main-page"
  >
    <div
      class="products"
    >
      <div
        class="productsActions"
      >
        <div
          class="filter"
        >
          <button
            class="filterButton _active"
            type="button"
          >
            Все
          </button>
          <button
            class="filterButton "
            type="button"
          >
            Мясные
          </button>
          <button
            class="filterButton "
            type="button"
          >
            Сырные
          </button>
          <button
            class="filterButton "
            type="button"
          >
            Вегетарианские
          </button>
          <button
            class="filterButton "
            type="button"
          >
            Острые
          </button>
        </div>
        <div
          class="sorter"
        >
          <span
            class="sorterTitle"
          >
            Сортировка по:
          </span>
          <div
            class="sorter-select css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-4-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class="sorter-select__control css-1s2u09g-control"
            >
              <div
                class="sorter-select__value-container sorter-select__value-container--has-value css-319lph-ValueContainer"
              >
                <div
                  class="sorter-select__single-value css-qc6sy-singleValue"
                >
                  цене (макс)
                </div>
                <input
                  aria-autocomplete="list"
                  aria-controls="react-select-4-listbox"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-owns="react-select-4-listbox"
                  aria-readonly="true"
                  class="css-mohuvp-dummyInput-DummyInput"
                  id="react-select-4-input"
                  inputmode="none"
                  role="combobox"
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class="sorter-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="sorter-select__indicator-separator css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="sorter-select__indicator sorter-select__dropdown-indicator css-tlfecz-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <p
        class="productsTitle"
      >
        Пицца
      </p>
      <div
        class="productsContent"
      >
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/mexican.jpg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Мексиканская
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                699
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/four-cheese.jpeg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Четыре сыра
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                699
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/pepperoni.jpeg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Пепперони
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                639
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/cheese-chiken.jpg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Сырный цыпленок
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                639
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/veg-mush.jpeg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Овощи и грибы
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                639
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/ham-cheese.jpeg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Ветчина и грибы
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                539
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/pizza-pie.jpg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Пицца-пирог
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                539
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/cheese.jpeg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Сырная U+1F331
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                449
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/double-chicken.jpeg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Двойной цыпленок
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                449
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
        <div
          class="pizza"
          data-testid="pizza-item"
        >
          <div
            class="pictureWrapper"
          >
            <img
              alt="pizza"
              class="picture"
              src="/React-Pizza/images/pepperoni-fresh.jpeg"
            />
          </div>
          <div
            class="content"
          >
            <p
              class="title"
            >
              Пепперони фреш
            </p>
            <div
              class="parameters"
            >
              <ul
                class="dough"
              >
                <li
                  class="
                  doughItem
                  
                  
               "
                >
                  тонкое
                </li>
                <li
                  class="
                  doughItem
                  _active
                  
               "
                >
                  традиционное
                </li>
              </ul>
              <ul
                class="size"
              >
                <li
                  class="sizeItem "
                >
                  25 см.
                </li>
                <li
                  class="sizeItem _active"
                >
                  30 см.
                </li>
                <li
                  class="sizeItem "
                >
                  35 см.
                </li>
              </ul>
            </div>
            <div
              class="order"
            >
              <p
                class="orderPrice"
              >
                449
                 
                <span>
                  ₽
                </span>
              </p>
              <button
                class="orderButton "
              >
                Добавить
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
